src/jaxoplanet/experimental/starry/pijk.py

```diff
@@ -0,0 +1,140 @@
+from typing import Any
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax.experimental.sparse import BCOO
+
+from jaxoplanet.types import Array
+
+
+class Pijk(eqx.Module):
+    """A class to represent and manipulate spherical harmonics in the polynomial basis.
+    Several indices are used throughout the class:
+    - Indices (i, j, k) represent the order of the polynomials of (x, y, z), for example
+    (1, 0, 2) represents x * z^2.
+    - Indices (l, m) represent the orders of the spherical harmonics.
+    - Index n represent the index of the polynomial in the flattened array.
+
+    Flattened array `to_dense` and `from_dense` follow th convention from Luger et al. (2019).
+    More specifically:
+
+    .. math::
+
+        \tilde{p} = (1, x, y, z, x^2, xz, xy, yz, y^2, ...)^T
+
+    """
+
+    data: dict[tuple[int, int, int], Array]
+    degree: int = eqx.field(static=True)
+    diagonal: bool = eqx.field(static=True)
+
+    def __init__(self, data: Mapping[tuple[int, int, int], Array]):
+        self.data = dict(data)
+        self.degree = max(self.ijk2lm(*ijk)[0] for ijk in data.keys())
+        self.diagonal = all(self.ijk2lm(*ijk)[1] == 0.0 for ijk in data.keys())
+
+    @property
+    def shape(self):
+        return (self.degree + 1) ** 2
+
+    @property
+    def indices(self):
+        return self.data.keys()
+
+    @staticmethod
+    def ijk2lm(i, j, k):
+        assert k in (0, 1)
+
+        if k == 0:
+            nu = 2 * j
+            mu = 2 * i
+        else:
+            nu = 2 * j + 1
+            mu = 2 * i + 1
+
+        m = int((nu - mu) / 2)
+        l = int((nu + mu) / 2)
+
+        return l, m
+
+    @staticmethod
+    def n2lm(n):
+        l = int(np.floor(np.sqrt(n)))
+        m = n - l**2 - l
+        return l, m
+
+    @staticmethod
+    def lm2n(l, m):
+        return l**2 + l + m
+
+    @staticmethod
+    def lm2ijk(l, m):
+        mu = l - m
+        nu = l + m
+
+        if nu % 2 == 0:
+            return mu // 2, nu // 2, 0
+        else:
+            return (mu - 1) // 2, (nu - 1) // 2, 1
+
+    @staticmethod
+    def n2ijk(n):
+        return Pijk.lm2ijk(*Pijk.n2lm(n))
+
+    def index(self, i, j, k):
+        l, m = self.ijk2lm(i, j, k)
+        return int((l + 1) * l + m)
+
+    def tosparse(self) -> BCOO:
+        indices, values = zip(*self.data.items())
+        idx = np.array([self.index(i, j, k) for i, j, k in indices])[:, None]
+        return BCOO((jnp.asarray(values), idx), shape=(self.shape,))
+
+    def todense(self) -> Array:
+        return self.tosparse().todense()
+
+    @classmethod
+    def from_dense(cls, x: Array, degree: int = None) -> "Pijk":
+        data = defaultdict(float)
+        if degree is None:
+            for i in range(len(x)):
+                data[Pijk.n2ijk(i)] = x[i]
+        else:
+            n = (degree + 1) ** 2
+            for i in range(n):
+                data[Pijk.n2ijk(i)] = x[i] if i < len(x) else 0.0
```

Comment:
We probably don't need to loop all the way up to `n`, since all of the higher order entries are `0.0` because of this conditional. We can just loop up to `len(x)`!

Reply:
True but we then need to set the `degree` of the map and modify the `toparse` function to output an array consistent with that degree. This is important for certain operations in the `maps.Map.render` and `light_curves.light_curve` functions. I think the indices in `toparse` should be computed from the degree anyway instead of taken from `data`. What do you think?